overlord/devicestate/handlers_systems.go

```diff
@@ -121,6 +132,148 @@ func purgeNewSystemSnapFiles(logfile string) error {
 	return s.Err()
 }
 
+func snapsUniqueToRecoverySystem(target string, systems []*System) ([]string, error) {
+	// asserted snaps are shared by systems, figure out which ones are unique to
+	// the system we want to remove
+	requiredByOtherSystems := make(map[string]bool)
+	for _, sys := range systems {
+		if sys.Label == target {
+			continue
+		}
+
+		sd, err := seed.Open(dirs.SnapSeedDir, sys.Label)
+		if err != nil {
+			return nil, err
+		}
+
+		if err := sd.LoadAssertions(nil, func(*asserts.Batch) error {
+			return nil
+		}); err != nil {
+			return nil, err
+		}
+
+		if err := sd.LoadMeta(seed.AllModes, nil, timings.New(nil)); err != nil {
+			return nil, err
+		}
+
+		err = sd.Iter(func(sn *seed.Snap) error {
+			if sn.ID() != "" {
+				requiredByOtherSystems[sn.Path] = true
+			}
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	targetSeed, err := seed.Open(dirs.SnapSeedDir, target)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := targetSeed.LoadAssertions(nil, func(*asserts.Batch) error {
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	if err := targetSeed.LoadMeta(seed.AllModes, nil, timings.New(nil)); err != nil {
+		return nil, err
+	}
+
+	var uniqueToTarget []string
+	err = targetSeed.Iter(func(sn *seed.Snap) error {
+		if sn.ID() != "" && !requiredByOtherSystems[sn.Path] {
+			uniqueToTarget = append(uniqueToTarget, sn.Path)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return uniqueToTarget, nil
+}
+
+func (m *DeviceManager) doRemoveRecoverySystem(t *state.Task, _ *tomb.Tomb) error {
+	systems, err := m.Systems()
```

Comment:
is this taking the lock on its own? it's strange to take the lock again just below, as things can change again in between? I suppose because of the conflict check that's not the case